### ISSUE TITLE
try/catch (in some cases) causes resolveCallIds to fail

### ISSUE THREAD

**Issue by Domiii** [Domiii/dbux#389]
## Reproduction

* run express#20 ("should throw when Content-Type is an array")
  * you see several errors coming from `RuntimeDataProvider.resolveCallIds`
    * NOTE: this tries to look up the `callId` (traceId) of any `CallExpression` that was recorded while running; based on `staticTrace` data
  * the issue seems to come from the fact that an error is thrown inside the `try` block in `lib\router\layer.js:95`
* express#18 shows similar symptoms.

NOTE: For some reason `samples/error1.js` does not suffer from that problem.
* -> Maybe it only happens from nested function calls inside the `try` block? Not sure...

## Steps to take

* [x] Talk to @MiccWan  (he is already researching this problem)
* [x] First, find out if the instrumentation data is complete: i.e. every call-related trace must have a `callId` and the call result trace must have a `resultCallId`.
  * -> Maybe add a Verbose mode that monitors the call stack resolution algorithm (similar to how Verbose mode in `traceVisitors` monitors the AST exploration)
* [x] There seems to be a bug where `ThrowArgument` has **two** result traces (but only one BCE). This might or might not be related (but also needs fixing).
  * At least this is what happens in Express#8.
  * Need to isolate and fix this in instrumentation.
  * NOTE: Traced in #346

**Comment by MiccWan** [comment]
#346 Has mentioned that ThrowArgument is traced more than once.

**Comment by MiccWan** [comment]
## Status
In PR #459 , I've fix some bug in algorithm of `resolveCallIds`, which makes `express#18` and `express#20` runnable.

## Problems
However, they get error again when reload them from log file, since `resolveErrorTraces` will change `trace.staticTraceId`(this time it resolves callIds with new staticTraceId).
In PR #457, I've disabled `Collection.postAdd` on deserialize, so this should be fixed after merge the two PR, but maybe we still have to figure out "why can't we resolve callIds from new staticTraceId".

**Comment by Domiii** [comment]
As we discussed, there are more inconsistencies in the way we are dealing with new data. In `postAdd`, we do two different things at different times:

1. Change actual data
   * For this, we want to pass an `isRaw` (instead of `applyPostAdd`) parameter for decision making in `Collection.postAdd`.
   * Maybe, we can split `postAdd` into two: `postAddRaw` (only called if `isRaw`) and `postAddProcessed` (always called)?
2. Maintain internal data structures
   * As it turns out, this is only used in `PDP`. We can replace all those cases with a slightly generalized `CollectionIndex` (see code below).
3. `PDP` and `RDP` are also not quite consistent in the way that `deserialize` is implemented. Can we bring them closer together?

# New CollectionIndex idea:

```js
import { newLogger } from '@dbux/common/src/log/logger';
import EmptyObject from '@dbux/common/src/util/EmptyObject';


const ArrayIndexManager = {
  createNew() {
    return [];
  },

  getAllKeys() {
    // ...
  }
};

const MapIndexManager = {
  // ...
  createNew() {
    return new Map();
  },

  // TODO
};

/**
 * In order to support non-array containers per key, override:
 * `get`, `getFirst`, `addEntryToKey`, `createNewContainerForKey`
 * 
 * @template {T}
 */
export default class CollectionIndex {
  /**
   * @type {DataProvider}
   */
  dp;
  name;

  /**
   * @type {T[][]}
   */
  _byKey;

  constructor(collectionName, indexName, { addOnNewData = true, stringKey = false } = EmptyObject) {
    this.collectionName = collectionName;
    this.name = indexName;
    this.log = newLogger(`${indexName} (Index)`);
    this.addOnNewData = addOnNewData;
    this.stringKey = stringKey;

    // TODO: change `stringKey` to `isMap` (instead of array)
    if (this.stringKey) {
      // this._manager = ...
    }
    else {
      // this._manager = ...
    }
  }

  _init(dp) {
    this.dp = dp;
    if (this.stringKey) {
      this._byKey = {};
    }
    else {
      this._byKey = [];
    }
  }

  /**
   * @param {number} key 
   * @return {T[]}
   */
  get(key) {
    return this._byKey[key];
  }

  getFirst(key) {
    const ofKey = this.get(key);
    return ofKey?.[0] || null;
  }

  getAllKeys() {
    const keys = [];
    for (let i = 0; i < this._byKey.length; ++i) {
      if (this._byKey[i]) {
        keys.push(i);
      }
    }
    return keys;
  }

  /**
   * @param {T[]} entries 
   */
  addEntries(entries) {
    for (const entry of entries) {
      this.addEntry(entry);
    }
  }

  /**
   * @param {T} entry 
   */
  addEntry(entry) {
    const key = this.makeKey(this.dp, entry);
    if (key === undefined) {
      // debugger;
      this.log.error('makeKey returned undefined');
      return;
    }
    if (key === false) {
      // entry is filtered out; not part of this index
      return;
    }

    // TODO: fix `this._byKey[key]`. Use manager instead.
    const ofKey = (this._byKey[key] = this._byKey[key] || this.createNewContainerForKey(key));
    this.beforeAdd?.(ofKey, entry);
    this.addEntryToKey(ofKey, entry);

    // sanity check
    // if (ofKey.includes(undefined)) {
    //   this.log.error('Index contains undefined values', key, entry, ofKey);
    // }
  }

  createNewContainerForKey(key) {
    return [];
  }

  addEntryToKey(ofKey, entry) {
    ofKey.push(entry);
  }

  addEntryById(id) {
    const entry = this.dp.collections[this.collectionName].getById(id);
    this.addEntry(entry);
  }

  /** 
   * Returns a unique key (number) for given entry.
   * @param {DataProvider} dp
   * @param {T} entry 
   */
  makeKey(/* dp, entry */) {
    throw new Error(`abstract method not implemented: ${this.constructor.name}.makeKey`);
  }
}
```

**Comment by MiccWan** [comment]
Solved in PR #459 , note that `PDP` and `RDP` have different save demands(`PDP` needs to save right after new data comes in, while `RDP` only wants to save all data when needed), so `serialize/deserialize` and `save/load` can not be merged easily.

**Comment by Domiii** [comment]
Express#20 has long been fixed, and, while still dirty, `resolveCallIds` has been revised and its internal error reporting largely improved several times since.
Also, @MiccWan has done a great job, fixing up mentioned `RDP` and `PDP` issues.
